File: MMOCoreORB/src/server/zone/managers/jedi/JediManager.cpp
/*
				Copyright <SWGEmu>
		See file COPYING for copying conditions.*/

#include "JediManager.h"
#include "server/zone/managers/director/DirectorManager.h"

JediManager::JediManager() : Logger("JediManager") {
	jediProgressionType = NOJEDIPROGRESSION;
	setJediManagerName("JediManager");
}

JediManager::~JediManager() {

}

const String& JediManager::getJediManagerName() {
	ReadLocker locker(this);

	return jediManagerName;
}

int JediManager::getJediProgressionType() {
	ReadLocker locker(this);

	return jediProgressionType;
}

void JediManager::setJediManagerName(const String& name) {
	Locker writeLock(this);

	jediManagerName = name;
}

void JediManager::setupLuaValues(Lua* luaEngine) {
	luaEngine->setGlobalInt("NOJEDIPROGRESSION", JediManager::NOJEDIPROGRESSION);
	luaEngine->setGlobalInt("HOLOGRINDJEDIPROGRESSION", JediManager::HOLOGRINDJEDIPROGRESSION);
	luaEngine->setGlobalInt("VILLAGEJEDIPROGRESSION", JediManager::VILLAGEJEDIPROGRESSION);
	luaEngine->setGlobalInt("CUSTOMJEDIPROGRESSION", JediManager::CUSTOMJEDIPROGRESSION);
	luaEngine->setGlobalInt("ITEMHOLOCRON", JediManager::ITEMHOLOCRON);
	luaEngine->setGlobalInt("ITEMWAYPOINTDATAPAD", JediManager::ITEMWAYPOINTDATAPAD);
	luaEngine->setGlobalInt("ITEMTHEATERDATAPAD", JediManager::ITEMTHEATERDATAPAD);
}

void JediManager::loadConfiguration(Lua* luaEngine) {
	setupLuaValues(luaEngine);

	luaEngine->runFile("scripts/managers/jedi/jedi_manager.lua");

	jediProgressionType = luaEngine->getGlobalInt(String("jediProgressionType"));

	switch (jediProgressionType) {
	case HOLOGRINDJEDIPROGRESSION:
		luaEngine->runFile("scripts/managers/jedi/hologrind_jedi_manager.lua");
		break;
	case VILLAGEJEDIPROGRESSION:
		luaEngine->runFile("scripts/managers/jedi/village_jedi_manager.lua");
		break;
	case CUSTOMJEDIPROGRESSION:
		luaEngine->runFile(luaEngine->getGlobalString(String("customJediProgressionFile")));
		break;
	default:
		break;
	}

	setJediManagerName(luaEngine->getGlobalString(String("jediManagerName")));

	info("Loaded.");
}

void JediManager::onPlayerCreated(CreatureObject* creature) {
	Lua* lua = DirectorManager::instance()->getLuaInstance();
	Reference<LuaFunction*> luaOnPlayerCreated = lua->createFunction(getJediManagerName(), "onPlayerCreated", 0);
	*luaOnPlayerCreated << creature;

	luaOnPlayerCreated->callFunction();
}

void JediManager::onPlayerLoggedIn(CreatureObject* creature) {
	Lua* lua = DirectorManager::instance()->getLuaInstance();
	Reference<LuaFunction*> luaOnPlayerLoggedIn = lua->createFunction(getJediManagerName(), "onPlayerLoggedIn", 0);
	*luaOnPlayerLoggedIn << creature;

	luaOnPlayerLoggedIn->callFunction();
}

void JediManager::onPlayerLoggedOut(CreatureObject* creature) {
	Lua* lua = DirectorManager::instance()->getLuaInstance();
	Reference<LuaFunction*> luaOnPlayerLoggedOut = lua->createFunction(getJediManagerName(), "onPlayerLoggedOut", 0);
	*luaOnPlayerLoggedOut << creature;

	luaOnPlayerLoggedOut->callFunction();
}

void JediManager::checkForceStatusCommand(CreatureObject* creature) {
	Lua* lua = DirectorManager::instance()->getLuaInstance();
	Reference<LuaFunction*> luaCheckForceStatusCommand = lua->createFunction(getJediManagerName(), "checkForceStatusCommand", 0);
	*luaCheckForceStatusCommand << creature;

	luaCheckForceStatusCommand->callFunction();
}

void JediManager::useItem(SceneObject* item, const int itemType, CreatureObject* creature) {
	Lua* lua = DirectorManager::instance()->getLuaInstance();
	Reference<LuaFunction*> luaUseItem = lua->createFunction(getJediManagerName(), "useItem", 0);
	*luaUseItem << item;
	*luaUseItem << itemType;
	*luaUseItem << creature;

	luaUseItem->callFunction();
}

bool JediManager::canLearnSkill(CreatureObject* creature, const String& skillName) {
	Lua* lua = DirectorManager::instance()->getLuaInstance();
	Reference<LuaFunction*> luaStartTask = lua->createFunction(getJediManagerName(), "canLearnSkill", 1);
	*luaStartTask << creature;
	*luaStartTask << skillName;

	lua_State* L = luaStartTask->callFunction();

	bool result = lua_toboolean(L, -1);

	lua_pop(L, 1);

	return result;
}

bool JediManager::canSurrenderSkill(CreatureObject* creature, const String& skillName) {
	Lua* lua = DirectorManager::instance()->getLuaInstance();
	Reference<LuaFunction*> luaStartTask = lua->createFunction(getJediManagerName(), "canSurrenderSkill", 1);
	*luaStartTask << creature;
	*luaStartTask << skillName;

	lua_State* L = luaStartTask->callFunction();

	bool result = lua_toboolean(L, -1);

	lua_pop(L, 1);

	return result;
}

void JediManager::onFSTreeCompleted(CreatureObject* creature, const String& branch) {
	Lua* lua = DirectorManager::instance()->getLuaInstance();
	Reference<LuaFunction*> luaStartTask = lua->createFunction(getJediManagerName(), "onFSTreeCompleted", 0);
	*luaStartTask << creature;
	*luaStartTask << branch;

	luaStartTask->callFunction();
}

File: MMOCoreORB/bin/scripts/managers/jedi/jedi_manager.lua
JediManager = require("screenplays.screenplay")

if _TEST then
	NOJEDIPROGRESSION = 0
	HOLOGRINDJEDIPROGRESSION = 1
	VILLAGEJEDIPROGRESSION = 2
	CUSTOMJEDIPROGRESSION = 3

	ITEMHOLOCRON = 0
	ITEMWAYPOINTDATAPAD = 1
	ITEMTHEATERDATAPAD = 2
end

jediProgressionType = VILLAGEJEDIPROGRESSION

jediManagerName = "JediManager"

customJediProgressionFile = "scripts/managers/jedi/custom_jedi_manager.lua"

function printf(...) io.write(string.format(table.unpack({...}))) end

JediManager = ScreenPlay:new {
	screenplayName = jediManagerName,
	jediManagerName = jediManagerName,
	jediProgressionType = NOJEDIPROGRESSION,
}

-- Default handling of the start event.
function JediManager:start()
	-- Default behaviour for the start event, do nothing.
end

-- Default handling of the onPlayerCreated event.
-- @param pCreatureObject pointer to the creature object for the player that was created.
function JediManager:onPlayerCreated(pCreatureObject)
	-- Default behaviour for the onPlayerCreated event, do nothing.
end

-- Default handling of the onPlayerLoggedIn event.
-- @param pCreatureObject pointer to the creature object for the player that logged in.
function JediManager:onPlayerLoggedIn(pCreatureObject)
	-- Default behaviour for the onPlayerLoggedIn event, do nothing.
end

-- Default handling of the onPlayerLoggedOut event.
-- @param pCreatureObject pointer to the creature object for the player that logged out.
function JediManager:onPlayerLoggedOut(pCreatureObject)
	-- Default behaviour for the onPlayerLoggedOut event, do nothing.
end

-- Default handling of the checkForceStatus command.
-- @param pCreatureObject pointer to the creature object for the player that performed the command.
function JediManager:checkForceStatusCommand(pCreatureObject)
	-- Default behaviour for the checkForceStatus command, do nothing.
end

-- Default handling of the onSkillRevoked event.
-- @param pCreatureObject pointer to the creature object that revoked a skill
-- @param pSkill skill that was revoked
function JediManager:onSkillRevoked(pCreatureObject, pSkill)
	-- Default behaviour for the onPlayerLoggedOut event, do nothing.
end

-- Default handling of the useItem event.
-- @param pSceneObject pointer to the item object.
-- @param itemType the type of item that is used.
-- @param pCreatureObject pointer to the creature object that used the item.
function JediManager:useItem(pSceneObject, itemType, pCreatureObject)
	-- Default behaviour for the useItem event, do nothing.
end

--Check for force skill prerequisites
function JediManager:canLearnSkill(pPlayer, skillName)
	-- Default behaviour for the canLearnSkill check, return false.
	return false
end

--Check to ensure force skill prerequisites are maintained
function JediManager:canSurrenderSkill(pPlayer, skillName)
	-- Default behaviour for the canSurrenderSkill check, return true.
	return true
end

function JediManager:onFSTreeCompleted(pCreatureObject, branch)
	-- Default behaviour for the onFSTreesCompleted event, do nothing.
end

return JediManager

